File: 3_1_GraphTraversal_DFS(Stack).py
import sys
from queue import LifoQueue
sys.stdin = open('input.txt', 'r')
sys.stdout = open('output.txt', 'w')

def dfsOfGraph(nodes, start, adj):
    visited = [False]*nodes
    visited[start]=True
    stack=LifoQueue()  # Using a list as a stack (LIFO behavior)
    stack.put(start)
    dfs =[]
    
    while not stack.empty():
        node=stack.get()
        dfs.append(node)
        
        for it in adj[node]:
            if not visited[it]:
                visited[it]=True
                stack.put(it)
    return dfs
# Read input
n, m = map(int, input().split())
adj = [[] for _ in range(n)]  # 0-based index

for _ in range(m):
    u, v = map(int, input().split())
    adj[u].append(v)
    # adj[v].append(u)  # Uncomment if the graph is undirected

# Run DFS from node 0
list_result = dfsOfGraph(6, 0, adj)  # Same as C++ version
print(*list_result)  # Print result space-separated


#for input
# 6 9
# 0 1
# 1 2
# 2 3
# 3 4
# 4 5
# 0 2
# 1 3
# 4 1
# 4 0

File: 3_1_GraphTraversal_DFS(Stack).cpp
#include<iostream>
#include<vector>
#include<stack>
using namespace std;
vector<int>dfsOfGraph(int nodes,int start, vector<int>adj[])
{
    vector<int>visited(nodes,false);
    visited[start]=true;
    stack<int>s;
    s.push(start);
    vector<int>dfs;
    while(!s.empty())
    {
        int node=s.top();
        s.pop();
        dfs.push_back(node);
        for(int it:adj[node])
        {
            if(!visited[it])
            {
                visited[it]=1;
                s.push(it);
            }
        }
    }
    return dfs;
     
}
int main()
{
    ////////File I/O Setup////////
     
    #ifndef ONLINE_JUDGE
    freopen("input.txt","r",stdin);
    freopen("output.txt","w",stdout);
    #endif
    
    //////////////////////////////
    int n,m;
    cin>>n>>m;
    vector<int>adj[n+1];
    for(int i=0;i<m;i++)
    {
        int u,v;
        cin>>u>>v;
        adj[u].push_back(v);
        //adj[v].push_back(u); //for directed graph
    }
    vector<int>list=dfsOfGraph(6,0,adj);
    for(int it:list)cout<<it<<" ";

}
